### PR TITLE
SI-8519  collection.immutable.Map.apply is inefficient

### PR DESCRIPTION
**Pull request by Ichoran**
Added customized apply and contains methods for EmptyMap - Map4 and ListMap (both empty and Node).

### CODE REVIEW

**Comment by Ichoran** [comment]
Note: JIRA didn't let me log in, so I haven't linked to this PR from there.

**Comment by Ichoran** [comment]
The test passes on my machine...not sure what the issue is?

**Comment by Ichoran** [comment]
PLS REBUILD/pr-scala@693b0647f86e9702ac21e87f5e093346bf2d6edd

**Comment by retronym** [comment]
```
fail - run/t8549.scala
...
java.io.InvalidClassException: scala.collection.immutable.ListMap$EmptyListMap$; local class incompatible: stream classdesc serialVersionUID = -8256686706655863282, local class serialVersionUID = 307923064409500379
```

Your change might have changed the implied `serialVersionUID` of the module class `EmptyListMap$` . Perhaps this was only under `-optimize`, which might explain why if failed on Jenkins but not for you.

We might be able to retrofit `@serialVerionUID(-8256686706655863282L)` on that object to make things work.

I think the principled approach is to actually use the serialization proxy pattern to avoid serializing these `MapN` / `SetN` / ... in the first place. 

To help debugging, here's how to extract the SerialVersionUID (calculated or explicitly annotated) of a class programatically:
https://github.com/scala/scala/commit/f9abdcef6c9b8d96e1aaf98942938e2e875285d0#diff-899f0b1feda908d4ef15744fae0fb8dbR9

**Comment by Ichoran** [comment]
@retronym - Thanks, I should have looked into it more deeply to see why it was failing.  (Not the first time I have forgotten the difference between -optimize and not.)  I guess we're trying to be serialization compatible in 2.12, so the thing to do for now would be to apply the annotation.

**Comment by lrytz** [comment]
> To help debugging, here's how to extract the SerialVersionUID (calculated or explicitly annotated) of a class programatically:
> f9abdce#diff-899f0b1feda908d4ef15744fae0fb8dbR9

You can also use the [`serialver` command line tool](https://docs.oracle.com/javase/7/docs/technotes/tools/solaris/serialver.html)

File: src/library/scala/collection/immutable/ListMap.scala [hunk 1]
```diff
@@ -176,7 +179,16 @@ extends AbstractMap[A, B]
...
+      
+      
+    override def contains(key: A): Boolean = contains0(this, key)
+    
+    @tailrec private def contains0(cur: ListMap[A, B1], k: A): Boolean =
+      if (k == cur.key) true
+      else if (cur.next.nonEmpty) contains0(cur.next, k)
+      else false
+
+
```
Comment (lrytz) [inline src/library/scala/collection/immutable/ListMap.scala:191]: could eliminate the excess newline

**Comment by lrytz** [comment]
LGTM, otherwise

**Comment by retronym** [comment]
LGTM, too.

**Pull request merged**